sdk/tests/core/test_az_json.c

```diff
@@ -791,276 +618,271 @@ az_result write_str(az_span span, az_span s, az_span* out, int32_t* written)
   return AZ_OK;
 }
 
-/** Json pointer **/
-static void test_json_pointer(void** state)
+// Using a macro instead of a helper method to retain line number
+// in call stack to help debug which line/test case failed.
+#define test_json_parser_invalid_helper(json, expected_result) \
+  do \
+  { \
+    az_json_parser parser = { 0 }; \
+    TEST_EXPECT_SUCCESS(az_json_parser_init(&parser, json, NULL)); \
+    az_result result = AZ_OK; \
+    while (result == AZ_OK) \
+    { \
+      result = az_json_parser_move_to_next_token(&parser); \
+    } \
+    assert_int_equal(result, expected_result); \
+  } while (0)
+
+static void test_json_parser_invalid(void** state)
```

Comment:
I added a whole bunch of invalid JSON test cases by reasoning through the various states (note: invalid, not incomplete). Let me know if folks come up with some more cases.

cc @vhvb1989